tests/test_app/test_install_plugins.py

```diff
@@ -0,0 +1,256 @@
+#
+# Note: Due to the fact that it uses import_lib caches stevedore will see the first
+# installed app immediately, but subsequent apps may not be seen until the cache is
+# updated or the interpreter reloaded.
+
+
+import importlib
+import json
+import os
+import site
+import sys
+import time
+
+from contextlib import contextmanager, suppress
+from importlib.metadata import distributions
+from inspect import getsourcelines, isclass
+from os import path
+from shutil import rmtree
+from subprocess import check_call
+from tempfile import mkdtemp
+
+import pytest
+
+import cogent3
+
+from cogent3.app import apps, available_apps, get_app
+from cogent3.app.composable import define_app
+
+
+def is_package_installed(package_name):
+    """Check if a package is installed"""
+    site_packages_dir = site.getsitepackages()[1]
+    installed_packages = os.listdir(site_packages_dir)
+    return package_name in installed_packages
+
+
+def install_app(cls, temp_dir, mod=None):
+    """
+    Installs a temporary app created from a class, by writing the source code
+    into a new Python file in the temporary directory, creating a setup.py file
+    that references the new module, and running pip install . the temporary directory.
+    """
+    # Get the source code of the function
+    lines, _ = getsourcelines(cls)
+    source_code = "\n".join(lines)
+    if not mod:
+        mod = f"mod_{cls.__name__}"
+
+    # Write the source code into a new Python file in the temporary directory
+    with open(path.join(temp_dir, f"{mod}.py"), "w") as f:
+        f.write("from cogent3.app.composable import define_app\n")
+        if hasattr(cls, "_requires_imports"):
+            for import_statement in cls._requires_imports():
+                f.write(f"{import_statement}\n")
+        f.write(source_code)
+
+    # Create a setup.py file in the temporary directory that references the new module
+    setup_py = f"""
+from setuptools import setup
+
+setup(
+name='{mod}',
+version='0.1',
+py_modules=['{mod}'],
+entry_points={{
+    "cogent3.app": [
+        "{cls.__name__} = {mod}:{cls.__name__}",
+    ],
+}},    
+)
+"""
+    with open(path.join(temp_dir, "setup.py"), "w") as f:
+        f.write(setup_py)
+
+    # Use subprocess to run pip install . in the temporary directory
+    check_call([sys.executable, "-m", "pip", "install", "."], cwd=temp_dir)
+
+    # Wait for the installation to complete
+    timeout = 60  # maximum time to wait in seconds
+    start_time = time.time()
+
+    start_time = time.time()
+    package_name = mod
+
+    while True:
+        if is_package_installed(package_name=package_name):
+            break
+        elif time.time() - start_time > timeout:
+            raise TimeoutError(
+                f"Package {package_name!r} not found after waiting for {timeout} seconds."
+            )
+        time.sleep(1)
+
+    # check we can import it
+    timeout = 60  # maximum time to wait in seconds

```

Comment:
duplicate definition of timeout, and below too...

Reply:
cut and paste antipattern.  I have an idea for a more declarative approach